### ISSUE TITLE
[FIRRTL] Add Verifier Check that Register Init Type Matches Register Type

### ISSUE THREAD

**Issue by seldridge** [llvm/circt#1477]
Add a check that FIRRTL's `regreset` errors if the type of the reset value doesn't match the type of reset.  E.g., the following should fail because a vector register is being reset to a ground type.  However, it does not fail and will crash LowerTypes:

```mlir
%r = firrtl.regreset %clock, %reset, %c0_ui1  : (!firrtl.clock, !firrtl.asyncreset, !firrtl.uint<1>) -> !firrtl.vector<uint<1>, 1>
```

I think there's something clever that can be done here with the `AllTypesMatch` or `AllTypesMatchWith` ODS functions (or something similar). However, I'm not finding an immediate solution.

**Comment by lattner** [comment]
Agreed, this should be a verifier check.

**Comment by seldridge** [comment]
Fixed via 3af1bd7dc3e29127500a874013542ad5515143cc.
